### ISSUE TITLE
ResourceWriteSuccessData - Authorization / Claims should be DYNAMIC or objects instead of strings

### ISSUE THREAD

**Issue by knom** [Azure/azure-sdk-for-net#5189]
E.g. file https://github.com/Azure/azure-sdk-for-net/blob/40248f466aa063bd90f16dae10d06b9b7317e2d0/src/SDKs/EventGrid/DataPlane/Microsoft.Azure.EventGrid/Generated/Models/ResourceWriteSuccessData.cs

I'm running into JSON Serialization issues with the EventGrid JSON..

As AUTHORIZATION + CLAIMS are no longer string, but really a subobject!

**Comment by NickDrouin** [comment]
The following code repro's the bug:

`        string resourceWriteSuccessDataString = "{\r\n  \"authorization\": {\r\n    \"scope\": \"/subscriptions/00000000-1111-2222-3333-000000000002\",\r\n    \"action\": \"Microsoft.Resources/deployments/write\",\r\n    \"evidence\": {\r\n      \"role\": \"Subscription Admin\"\r\n    }\r\n  },\r\n  \"claims\": {\r\n    \"aud\": \"https://management.core.windows.net/\",\r\n    \"manyMore\": \"claimValues\",\r\n    \"ver\": \"1.0\"\r\n  },\r\n  \"correlationId\": \"00000000-1111-2222-3333-000000000004\",\r\n  \"resourceProvider\": \"Microsoft.Resources\",\r\n  \"resourceUri\": \"/subscriptions/00000000-1111-2222-3333-000000000002/providers/Microsoft.Resources/deployments/00000000-1111-2222-3333-000000000001-hsf-initial\",\r\n  \"operationName\": \"Microsoft.Resources/deployments/write\",\r\n  \"status\": \"Succeeded\",\r\n  \"subscriptionId\": \"00000000-1111-2222-3333-000000000002\",\r\n  \"tenantId\": \"00000000-1111-2222-3333-000000000003\"\r\n}";
        

[TestMethod]
        public void ResourceWriteSuccessDataWorkaroundStillNeeded()
        {
            Assert.ThrowsException<JsonReaderException>(() =>
                JsonConvert.DeserializeObject<ResourceWriteSuccessData>(resourceWriteSuccessDataString)
            );
            // {"Unexpected character encountered while parsing value: {. Path 'authorization', line 2, position 20."}
        }`

Using libs: 
netcoreapp2.2
"Microsoft.AspNetCore.Mvc" Version="2.2.0"
"Microsoft.Azure.EventGrid" Version="3.2.0"
"Newtonsoft.Json" Version="12.0.2"

To work-around, I copied the class and changed the authorization and claims to 'object' instead of 'string'.

**Comment by NickDrouin** [comment]
@ahamad-MS, here are some xunit tests, which can be added to:

https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/eventgrid/Microsoft.Azure.EventGrid/tests/Tests/ConsumeEventTests.cs#L853

Once the bug is resolved.

```
        [Fact]
        public void ConsumeResourceWriteSuccessEventWithClaimsJObject()
        {
            string requestContent = "[   {     \"topic\":\"/subscriptions/{subscription-id}\",     \"subject\":\"/subscriptions/{subscription-id}/resourceGroups/{resource-group}/providers/Microsoft.EventGrid/eventSubscriptions/LogicAppdd584bdf-8347-49c9-b9a9-d1f980783501\",    \"eventType\":\"Microsoft.Resources.ResourceWriteSuccess\",    \"eventTime\":\"2017-08-16T03:54:38.2696833Z\",    \"id\":\"25b3b0d0-d79b-44d5-9963-440d4e6a9bba\",    \"data\": {        \"authorization\":\"{azure_resource_manager_authorizations}\",        \"claims\": {\"someClaim\":\"theClaimValue\", \"someOtherClaim\":\"theOtherClaimValue\"},        \"correlationId\":\"54ef1e39-6a82-44b3-abc1-bdeb6ce4d3c6\",        \"httpRequest\":\"{request-operation}\",        \"resourceProvider\":\"Microsoft.EventGrid\",        \"resourceUri\":\"/subscriptions/{subscription-id}/resourceGroups/{resource-group}/providers/Microsoft.EventGrid/eventSubscriptions/LogicAppdd584bdf-8347-49c9-b9a9-d1f980783501\",        \"operationName\":\"Microsoft.EventGrid/eventSubscriptions/write\",        \"status\":\"Succeeded\",        \"subscriptionId\":\"{subscription-id}\",        \"tenantId\":\"72f988bf-86f1-41af-91ab-2d7cd011db47\"        },      \"dataVersion\": \"\",      \"metadataVersion\": \"1\"  }]";

            var events = this.eventGridSubscriber.DeserializeEventGridEvents(requestContent);

            Assert.NotNull(events);
            Assert.True(events[0].Data is ResourceWriteSuccessData);
            ResourceWriteSuccessData eventData = (ResourceWriteSuccessData)events[0].Data;
            Assert.Equal("72f988bf-86f1-41af-91ab-2d7cd011db47", eventData.TenantId);
        }

        [Fact]
        public void ConsumeResourceWriteSuccessEventWithAuthorizationJObject()
        {
            string requestContent = "[   {     \"topic\":\"/subscriptions/{subscription-id}\",     \"subject\":\"/subscriptions/{subscription-id}/resourceGroups/{resource-group}/providers/Microsoft.EventGrid/eventSubscriptions/LogicAppdd584bdf-8347-49c9-b9a9-d1f980783501\",    \"eventType\":\"Microsoft.Resources.ResourceWriteSuccess\",    \"eventTime\":\"2017-08-16T03:54:38.2696833Z\",    \"id\":\"25b3b0d0-d79b-44d5-9963-440d4e6a9bba\",    \"data\": {        \"authorization\": { \"scope\": \"/subscriptions/00000000-1111-2222-3333-000000000002\", \"action\": \"Microsoft.Resources/deployments/write\", \"evidence\": { \"role\": \"Subscription Admin\" } },        \"claims\":\"{azure_resource_manager_claims}\",        \"correlationId\":\"54ef1e39-6a82-44b3-abc1-bdeb6ce4d3c6\",        \"httpRequest\":\"{request-operation}\",        \"resourceProvider\":\"Microsoft.EventGrid\",        \"resourceUri\":\"/subscriptions/{subscription-id}/resourceGroups/{resource-group}/providers/Microsoft.EventGrid/eventSubscriptions/LogicAppdd584bdf-8347-49c9-b9a9-d1f980783501\",        \"operationName\":\"Microsoft.EventGrid/eventSubscriptions/write\",        \"status\":\"Succeeded\",        \"subscriptionId\":\"{subscription-id}\",        \"tenantId\":\"72f988bf-86f1-41af-91ab-2d7cd011db47\"        },      \"dataVersion\": \"\",      \"metadataVersion\": \"1\"  }]";

            var events = this.eventGridSubscriber.DeserializeEventGridEvents(requestContent);

            Assert.NotNull(events);
            Assert.True(events[0].Data is ResourceWriteSuccessData);
            ResourceWriteSuccessData eventData = (ResourceWriteSuccessData)events[0].Data;
            Assert.Equal("72f988bf-86f1-41af-91ab-2d7cd011db47", eventData.TenantId);
        }
```

Note that I can't actually repro a ResourceWriteFailure event, but I'd wager the Authorization and Claims are similarly complex, and will fail in the same way.

To repro in azure, simply register your endpoint to the subscription (or resource group) for event grid events.

Impact Statement:
Affects deserialization as described above.
Affects EventGridSubscriber.DeserializeEventGridEvents function which fails in the same way.
Affects current MSFT partner engagements.

Regression: Unknown

**Comment by NickDrouin** [comment]
Looks like httpRequest is also a JObject now.

**Comment by ahamad-MS** [comment]
We are adding this for investigation in our backlog and will investigate/triage appropriately

**Comment by crgarcia12** [comment]
+1

**Comment by JoshLove-msft** [comment]
This is an issue in the new Azure.Messaging.EventGrid library as well.
I see two potential paths forward here:
 - Store these `object` values as JSON strings in the existing properties. You should be able to reparse to get it back as a JSON object.
 - Create new events with the correctly typed properties, and hide the old ones (to avoid breaking anyone).

**Comment by JoshLove-msft** [comment]
We fixed this by converting the JSON objects into strings and storing in the existing properties. We still want to add new Dictionary properties that would give you the JSON object. Reopening to track that work.

**Comment by JoshLove-msft** [comment]
This was fixed in https://github.com/Azure/azure-sdk-for-net/pull/26222 and is available as of 4.8.1.
